llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -5661,6 +5661,9 @@ InstructionCost AArch64TTIImpl::getPartialReductionCost(
       AccumType->getScalarSizeInBits() / InputTypeA->getScalarSizeInBits();
   if (VF.getKnownMinValue() <= Ratio)
     return Invalid;
+  // i32 -> i64 or i16 -> i32 is not natively supported on Neon and SVE.
+  if (Ratio < 4)
```

Comment:
A partial reduction from i32 -> i64 or i16 -> i32 is supported with `smlalb`+`smlalt`. If the cost at the end of the function would be slightly tweaked to `return Cost + 2;` (instead of `return Cost + 4`), then this loop vectorises with partial reductions and VF=4.